### PR TITLE
Allow concatenating two or more dirs for cores

### PR DESCRIPTION
**Pull request by fr500**
**Goals:**

* allow RetroArch to load cores from more than one location
* allow RetroArch to use the core updater and the distro package manager at the same time
* maybe allow RetroArch to have nightly cores and stable cores, or nightly cores, and netplay cores or whatever.
* allow lakka to use the updater!

**Description:**

This is particularly useful for *nixes since the package managers often install cores in a privileged location (/usr/lib/libretro for instance) (the same would work for windows if we ever embrace installing in Program Files).

Basically it allows to specify a secondary core direcotory in the configuration and concatenates both directories when creating core_info.

It's not 100% ready, it doesn't have a menu entry, and the "Load Core" menu doesn't use this yet, but the Suggested Cores does. To test it set libretro_directory_user to the user dir, and libretro_directory to the distro default location. It doesn't distinguish one or the other, we could use sublabels for that at the very least.

screenshot 2017-08-22 22 54 05

To leverage this, packagers would do the following:

* set the libretro_directory to "/usr/lib/libretro"
* we should make libretro_directory_user use the setting from g_defaults

Then the user could use both the updater, and the distro packages at the same time. It would also allow us to have stable and nightly cores sometime in the future

Alternatively we could rename the new dir to libretro_directory_distro, so we can leave the current dir alone, and set a default for that one instead. **Either way this will need work from the packager and the platform driver "mantainer"**

Let me know if this is interesting or acceptable, if so I can continue working on it

@twinaphex @hunterk @JuanVCS @sergiobenrocha2 @Kivutar

### CODE REVIEW

**Comment by hizzlekizzle** [comment]
I think this is an excellent addition. Being able to play nicely with package managers without sacrificing functionality is a good thing!

**Comment by twinaphex** [comment]
Sure, go ahead and continue wih it. If this means sergio-br2 no longer has to make a custom version then it would be positive.

**Comment by RobLoach** [comment]
Thanks for starting this up. Lakka had been using a unison mount to solve this, `directory_libretro_user` looks like a great solution!

**Comment by orbea** [comment]
This would be very much appreciated and would be useful for the RetroArch build script at SlackBuilds.org so users could either use the build bot or the libretro core build scripts at SBo. Here is a related issue.
https://github.com/libretro/RetroArch/issues/3237

I suggest that `libretro_directory` should be a configure option while `libretro_directory_user` should be set in the configuration files as it is currently.

**Comment by orbea** [comment]
You might also want to consider looking at duplicating this for core info files. Versions of cores could differ where a single info file may contain conflicting information.

**Comment by fr500** [comment]
The core info files don't really provide any accurate version information though and it adds a lot of complication to the mix

With configure option you mean a ./configure option?

**Comment by orbea** [comment]
Yes I mean a `./configure` option.

Since SBo only provides build scripts and not binary packages this gives the user a great deal of freedom in building RetroArch. I allow users to control it like this currently.
```
if [ "${BUILDBOT:-0}" != 1 ]; then
  SED_CORE="s|# libretro_directory =|libretro_directory = /usr/lib${LIBDIRSUFFIX}/libretro|"
  SED_INFO="s|# libretro_info_path =|libretro_info_path = /usr/lib${LIBDIRSUFFIX}/libretro/info|"
  SED_MENU='s|# menu_show_core_updater = true|menu_show_core_updater = false|'
else
  SED_CORE=
  SED_INFO=
  SED_MENU=
fi

# Set the config file default directories to be consistent with the installation.
sed -e "s|# audio_filter_dir =|audio_filter_dir = /usr/lib${LIBDIRSUFFIX}/retroarch/filters/audio|" \
       -e "s|# video_filter_dir =|video_filter_dir = /usr/lib${LIBDIRSUFFIX}/retroarch/filters/video|" \
       -e "$SED_CORE;$SED_INFO;$SED_MENU" \
    -i retroarch.cfg
```
So if RA is built by default the script will use sed on `retroarch.cfg` to change the `libretro_directory` and `libretro_info_path`directories to `/usr/lib${LIBDIRSUFFIX}/libretro` and `/usr/lib${LIBDIRSUFFIX}/libretro/info` as well as set `menu_show_core_updater` to false since the core and info online updates won't work. The info files would be supplied by each individual libretro core build script.

The alternative is when the user builds with `BUILDBOT=1` where it will simply skip these and use the build bot which will of course not be compatible with the slackware libretro build scripts. I also change the `audio_filter_dir` and `video_filter_dir`, but that is another issue.

**Comment by fr500** [comment]
I'll think about it.

Last night I thought of a better solution that would make things even more interesting.
Instead of concatenating core_info_list here:

https://github.com/libretro/RetroArch/pull/5365/files#diff-f9289feeb00489732328764ea00aa3b4R606

I should write a a generic dir_list_concatenate function.
That way I can just pass a concatenated list to the core_info stuff and it should "just work", the same for the basic core browser.

And it could be used for other interesting things:

- Provide just a basic overlay set in the distro bundle (or on the android apk) and overlay the core downloader one on top of that
Same applies for shaders, autoconf, etc.

- Have more than one content dir in "Start Directory", have them all show up as a single dir.

I shall test this new approach this weekend.

**Comment by RobLoach** [comment]
While I'm not quite sure it's the approach we'd want to take, another possible solution would be to use [PhysFS](https://www.icculus.org/physfs/). PhysFS allows you to "mount" different directories and archives, and then read files from any of those mount points.

It would look something like this...

``` c
// Initialize PhysFS
PHYSFS_init(NULL);

// Mount all the possible core paths
PHYSFS_AddToSearchPath(libretro_core_dir, 1);
PHYSFS_AddToSearchPath(libretro_core_user_dir, 1);

// See if the core file exists in any of the mount paths
if (PHYSFS_exists(coretoload)) {

  // Load the file.
  PHYSFS_file* myfile = PHYSFS_openRead(coretoload);

  // TODO: Load the core through myfile.

  // Finish up.
  PHYSFS_close(myfile);
}
```

PhysicsFS 3.0 is also [now in beta](https://www.patreon.com/posts/physicsfs-3-0-in-13866851).

**Comment by ofry** [comment]
Wow! Maybe, that can be used in VFS too?

**Comment by fr500** [comment]
We really don't need a dependency to show the files as if it was a single filesystem though.
Not sure about VFS, this is not about the VFS.

**Comment by ofry** [comment]
https://github.com/libretro/RetroArch/pull/3715

It can handle VFS too. (load smth from archives, join multiple directories, etc)

**Comment by ofry** [comment]
So why you against this PhysFS dependency? It already ported to many systems...

**Comment by fr500** [comment]
Where did I say I was against it?

**Comment by fr500** [comment]
Ok instead of the original commit which was core directory specific, I made a dir_list_concatenate function instead, so this could be used with minor effort to overlay two directories for shaders, or info, or cores, or cheats or anything really, even content.

To use it for the Load Core menu I basically did this:

File: menu/menu_displaylist.c [hunk 1]
```diff
@@ -3621,8 +3621,19 @@ static int menu_displaylist_parse_cores(
...
+   else
+   {
+      struct string_list *list1 = dir_list_new(info->path, info->exts, 
+                                    true, settings->bools.show_hidden_files, true, false);
+      struct string_list *list2 = dir_list_new(settings->paths.directory_libretro_distro, info->exts, 
+                                    true, settings->bools.show_hidden_files, true, false);
+      str_list = dir_list_concatenate(list1, list2);
+      string_list_free(list1);
+      string_list_free(list2);
+   }
```
Comment (fr500) [inline menu/menu_displaylist.c]: basically this is all the code needed to concatenate two paths so it shows as one

**Comment by fr500** [comment]
@orbea Ok I think I added what you wanted
You can do this now

`./configure --with-platform_libretro_dir=/foo/bar`

And it will set 
libretro_directory_platform to that on the first run

I just unearthed a bug, if the libretro_dir or this dir are not valid dirs it will segfault, but it happened before too, it's a bug on dir_list_read and shouldn't affect this PR (should be fixed on master)

The only thing needed imho is a sublabel indicating the dir which I don't know how to implement.

**Comment by fr500** [comment]
I just unearthed a bug, if the libretro_dir or this dir are not valid dirs it will segfault, but it happened before too, it's a bug on dir_list_read and shouldn't affect this PR (should be fixed on master)

**Comment by fr500** [comment]
Aaaaaanyway, this is done as far as I'm concerned, it just needs people to use it. If anything arises let me know, moving on to something else.

**Comment by fr500** [comment]
Ok... no it doesn't work all that well.
If I add this to "Load Core" it will never find the cores since the file browser always use the parent dir.

menu_displaylist_pare_cores strips the basedir anyhow. But I managed to readd it. Thing is once that was done in my case it was trying to load d:\msys2\home\andres\retroarch\cores\c:\libretrocores\snes9x.dll when I tried loading from the secondary dir, and d:\msys2\home\andres\retroarch\cores\d:\msys2\home\andres\retroarch\cores\snes9x_libretro.dll when I used the normal dir.

Maybe we need another entry in **enum msg_file_type** that allows us to specify that the entry is fullpath, but then someone will need to figure how to add that behavior to the file browser.

I give up for now.

**Comment by fr500** [comment]
This code remains here:
https://github.com/fr500/RetroArch/tree/coredirs

In case someone is willing to cherry pick some stuff or try again

**Pull request closed without merging**